messaging/mqtt/subscriber.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package mqtt
+
+import (
+	"fmt"
+	"time"
+
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"github.com/gogo/protobuf/proto"
+	"github.com/mainflux/mainflux/errors"
+	log "github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/messaging"
+)
+
+var _ messaging.Publisher = (*publisher)(nil)
+
+var (
+	errSubscribe   = errors.New("failed to subscribe")
+	errUnsubscribe = errors.New("failed to unsubscribe")
+)
+
+type subscriber struct {
+	client  mqtt.Client
+	timeout time.Duration
+	logger  log.Logger
+}
+
+// NewSubscriber returns a new MQTT message subscriber.
+func NewSubscriber(address string, timeout time.Duration, logger log.Logger) (messaging.Subscriber, error) {
+	client, err := newClient(address, timeout)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := subscriber{
+		client:  client,
+		timeout: timeout,
+		logger:  logger,
+	}
+	return ret, nil
+}
+
+func (sub subscriber) Subscribe(topic string, handler messaging.MessageHandler) error {
+	token := sub.client.Subscribe(topic, qos, sub.mqttHandler(handler))
+	ok := token.WaitTimeout(sub.timeout)
+	if ok && token.Error() != nil {
+		return token.Error()
```

Comment:
if we return `token.Error()` on multiple places it would be good to wrap with another error so that exact place can be more easily detected

Reply:
I would leave error wrapping for the layer above MQTT Publisher or Subscriber. We only use it as a wrapper around MQTT client, so returning the real error is sufficient, IMHO. However, I'll improve the message of the errors we return if token is expired to be more specific ("failed to subscribe/unsubscribe due to timeout reached" is better than just "failed to subscribe/unsubscribe").